File: packages/@webreel/core/src/timeline.ts
import { writeFileSync } from "node:fs";
import type { Point, SoundEvent } from "./types.js";
import type { ZoomEvent } from "./autozoom.js";
import {
  TARGET_FPS,
  DEFAULT_CURSOR_SVG,
  DEFAULT_VIEWPORT_SIZE,
  OFFSCREEN_MARGIN,
  DEFAULT_CURSOR_SIZE,
  DEFAULT_HUD_THEME,
} from "./types.js";

interface CursorState {
  x: number;
  y: number;
  scale: number;
}

interface HudState {
  labels: string[];
}

interface FrameData {
  cursor: CursorState;
  hud: HudState | null;
}

export interface TimelineData {
  fps: number;
  width: number;
  height: number;
  zoom: number;
  theme: {
    cursorSvg: string;
    cursorSize: number;
    cursorHotspot: "top-left" | "center";
    hud: {
      background: string;
      color: string;
      fontSize: number;
      fontFamily: string;
      borderRadius: number;
      position: "top" | "bottom";
    };
  };
  frames: FrameData[];
  events: SoundEvent[];
  zoomEvents?: ZoomEvent[];
}

export class InteractionTimeline {
  private cursorPath: Point[] | null = null;
  private pathIndex = 0;
  private currentCursor: CursorState = {
    x: -OFFSCREEN_MARGIN,
    y: -OFFSCREEN_MARGIN,
    scale: 1,
  };
  private currentHud: HudState | null = null;
  private frames: FrameData[] = [];
  private events: SoundEvent[] = [];
  private frameCount = 0;
  private tickResolvers: Array<() => void> = [];
  private pathCompleteResolvers: Array<() => void> = [];

  private width: number;
  private height: number;
  private zoom: number;
  private fps: number;
  private cursorSvg: string;
  private cursorSize: number;
  private cursorHotspot: "top-left" | "center";
  private hudConfig: TimelineData["theme"]["hud"];

  constructor(
    width = DEFAULT_VIEWPORT_SIZE,
    height = DEFAULT_VIEWPORT_SIZE,
    options?: {
      zoom?: number;
      fps?: number;
      initialCursor?: { x: number; y: number };
      cursorSvg?: string;
      cursorSize?: number;
      cursorHotspot?: "top-left" | "center";
      hud?: Partial<TimelineData["theme"]["hud"]>;
      loadedFrames?: FrameData[];
      loadedEvents?: SoundEvent[];
    },
  ) {
    this.width = width;
    this.height = height;
    this.zoom = options?.zoom ?? 1;
    this.fps = options?.fps ?? TARGET_FPS;
    if (options?.initialCursor) {
      this.currentCursor = {
        x: options.initialCursor.x,
        y: options.initialCursor.y,
        scale: 1,
      };
    }
    this.cursorSvg = options?.cursorSvg ?? DEFAULT_CURSOR_SVG;
    this.cursorSize = options?.cursorSize ?? DEFAULT_CURSOR_SIZE;
    this.cursorHotspot = options?.cursorHotspot ?? "top-left";
    this.hudConfig = {
      background: options?.hud?.background ?? DEFAULT_HUD_THEME.background,
      color: options?.hud?.color ?? DEFAULT_HUD_THEME.color,
      fontSize: options?.hud?.fontSize ?? DEFAULT_HUD_THEME.fontSize,
      fontFamily: options?.hud?.fontFamily ?? DEFAULT_HUD_THEME.fontFamily,
      borderRadius: options?.hud?.borderRadius ?? DEFAULT_HUD_THEME.borderRadius,
      position: options?.hud?.position ?? DEFAULT_HUD_THEME.position,
    };
    if (options?.loadedFrames) {
      this.frames = options.loadedFrames;
      this.frameCount = options.loadedFrames.length;
    }
    if (options?.loadedEvents) {
      this.events = options.loadedEvents;
    }
  }

  setCursorPath(positions: Point[]): void {
    this.cursorPath = positions;
    this.pathIndex = 0;
  }

  setCursorScale(scale: number): void {
    this.currentCursor.scale = scale;
  }

  showHud(labels: string[]): void {
    this.currentHud = { labels };
  }

  hideHud(): void {
    this.currentHud = null;
  }

  addEvent(type: "click" | "key"): void {
    const timeMs = (this.frameCount / this.fps) * 1000;
    this.events.push({ type, timeMs });
  }

  waitForNextTick(): Promise<void> {
    return new Promise((resolve) => {
      this.tickResolvers.push(resolve);
    });
  }

  // Resolves when the current cursorPath is fully consumed. If no path is
  // active, resolves immediately. Used by animateMoveTo to fire the next
  // action (e.g. a click) exactly when the cursor arrives at its target —
  // regardless of capture rate, tickDuplicate cadence, or hardware speed.
  waitForPathComplete(): Promise<void> {
    if (this.cursorPath === null) return Promise.resolve();
    return new Promise((resolve) => {
      this.pathCompleteResolvers.push(resolve);
    });
  }

  private maybeResolvePathComplete(): void {
    if (this.cursorPath !== null || this.pathCompleteResolvers.length === 0) return;
    const resolvers = this.pathCompleteResolvers;
    this.pathCompleteResolvers = [];
    for (const resolve of resolvers) resolve();
  }

  tick(): void {
    if (this.cursorPath && this.pathIndex < this.cursorPath.length) {
      const p = this.cursorPath[this.pathIndex++];
      this.currentCursor.x = p.x;
      this.currentCursor.y = p.y;
      if (this.pathIndex >= this.cursorPath.length) {
        this.cursorPath = null;
      }
    }

    this.pushCurrentState();

    const resolvers = this.tickResolvers;
    this.tickResolvers = [];
    for (const resolve of resolvers) resolve();

    this.maybeResolvePathComplete();
  }

  tickDuplicate(): void {
    // Advance the cursor along the precomputed path on duplicate slots, so
    // every 60fps output frame shows a unique cursor position even when the
    // underlying screenshot is repeated. Without this, cursor motion runs at
    // the capture rate (~28fps) in a 60fps container, producing visible stutter.
    //
    // Critical: do NOT fire tickResolvers here — those gate action timing
    // (e.g. typeText inter-keystroke cadence via waitForNextTick) on REAL
    // captured frames. Only the real `tick()` resolves them.
    // `pathCompleteResolvers` DO fire here though — a path that exhausts on
    // a duplicate slot is just as arrived as one that exhausts on a real tick.
    if (this.cursorPath && this.pathIndex < this.cursorPath.length) {
      const p = this.cursorPath[this.pathIndex++];
      this.currentCursor.x = p.x;
      this.currentCursor.y = p.y;
      if (this.pathIndex >= this.cursorPath.length) {
        this.cursorPath = null;
      }
    }
    this.pushCurrentState();

    this.maybeResolvePathComplete();
  }

  private pushCurrentState(): void {
    this.frames.push({
      cursor: { ...this.currentCursor },
      hud: this.currentHud ? { labels: [...this.currentHud.labels] } : null,
    });
    this.frameCount++;
  }

  getEvents(): SoundEvent[] {
    return this.events;
  }

  getFrameCount(): number {
    return this.frameCount;
  }

  toJSON(): TimelineData {
    return {
      fps: this.fps,
      width: this.width,
      height: this.height,
      zoom: this.zoom,
      theme: {
        cursorSvg: this.cursorSvg,
        cursorSize: this.cursorSize,
        cursorHotspot: this.cursorHotspot,
        hud: this.hudConfig,
      },
      frames: this.frames,
      events: this.events,
    };
  }

  save(path: string): void {
    writeFileSync(path, JSON.stringify(this.toJSON()));
  }

  static load(json: TimelineData): InteractionTimeline {
    return new InteractionTimeline(json.width, json.height, {
      zoom: json.zoom,
      fps: json.fps,
      cursorSvg: json.theme.cursorSvg,
      cursorSize: json.theme.cursorSize,
      cursorHotspot: json.theme.cursorHotspot,
      hud: json.theme.hud,
      loadedFrames: json.frames,
      loadedEvents: json.events,
    });
  }
}

File: packages/@webreel/compositor/src/timeline.zig
const std = @import("std");
const cursor_motion = @import("cursor_motion.zig");

pub const EventType = enum {
    click,
    key,
};

pub const SoundEvent = struct {
    event_type: EventType,
    time_ms: f64,
};

pub const CursorState = struct {
    x: f64,
    y: f64,
    scale: f64,
};

pub const HudState = struct {
    labels: []const []const u8,
};

pub const FrameData = struct {
    cursor: CursorState,
    hud: ?HudState,
};

pub const Timeline = struct {
    allocator: std.mem.Allocator,
    cursor_path: ?[]const cursor_motion.Point = null,
    path_index: usize = 0,
    scale_path: ?[]f64 = null,
    scale_path_index: usize = 0,
    current_cursor: CursorState,
    current_hud: ?HudState = null,
    frames: std.ArrayList(FrameData) = .empty,
    events: std.ArrayList(SoundEvent) = .empty,
    frame_count: usize = 0,
    width: u32,
    height: u32,
    zoom: f64,
    fps: u32,

    pub fn init(allocator: std.mem.Allocator, width: u32, height: u32, fps: u32, zoom: f64, initial_cursor: ?CursorState) Timeline {
        return .{
            .allocator = allocator,
            .width = width,
            .height = height,
            .fps = fps,
            .zoom = zoom,
            .current_cursor = initial_cursor orelse CursorState{ .x = -40, .y = -40, .scale = 1.0 },
            .frames = .empty,
            .events = .empty,
        };
    }

    pub fn deinit(self: *Timeline) void {
        self.frames.deinit(self.allocator);
        self.events.deinit(self.allocator);
        if (self.scale_path) |sp| self.allocator.free(sp);
    }

    pub fn setCursorPath(self: *Timeline, positions: []const cursor_motion.Point) void {
        self.cursor_path = positions;
        self.path_index = 0;
    }

    pub fn setCursorScale(self: *Timeline, scale: f64) void {
        self.current_cursor.scale = scale;
    }

    pub fn setCursorScaleAnimated(self: *Timeline, target_scale: f64, frame_count: usize) void {
        const start_scale = self.current_cursor.scale;
        const steps = self.allocator.alloc(f64, frame_count) catch return;

        for (0..frame_count) |i| {
            const t = @as(f64, @floatFromInt(i + 1)) / @as(f64, @floatFromInt(frame_count));
            const eased = 1.0 - (1.0 - t) * (1.0 - t);
            steps[i] = start_scale + (target_scale - start_scale) * eased;
        }

        if (self.scale_path) |old| self.allocator.free(old);
        self.scale_path = steps;
        self.scale_path_index = 0;
    }

    pub fn showHud(self: *Timeline, labels: []const []const u8) void {
        self.current_hud = .{ .labels = labels };
    }

    pub fn hideHud(self: *Timeline) void {
        self.current_hud = null;
    }

    pub fn addEvent(self: *Timeline, event_type: EventType) void {
        const time_ms = @as(f64, @floatFromInt(self.frame_count)) / @as(f64, @floatFromInt(self.fps)) * 1000.0;
        self.events.append(self.allocator, .{ .event_type = event_type, .time_ms = time_ms }) catch {};
    }

    pub fn tick(self: *Timeline) void {
        if (self.cursor_path) |path| {
            if (self.path_index < path.len) {
                const p = path[self.path_index];
                self.current_cursor.x = p.x;
                self.current_cursor.y = p.y;
                self.path_index += 1;
                if (self.path_index >= path.len) {
                    self.cursor_path = null;
                }
            }
        }

        if (self.scale_path) |sp| {
            if (self.scale_path_index < sp.len) {
                self.current_cursor.scale = sp[self.scale_path_index];
                self.scale_path_index += 1;
                if (self.scale_path_index >= sp.len) {
                    self.allocator.free(sp);
                    self.scale_path = null;
                }
            }
        }

        self.pushCurrentState();
    }

    pub fn getLastFrame(self: *const Timeline) ?FrameData {
        if (self.frames.items.len == 0) return null;
        return self.frames.items[self.frames.items.len - 1];
    }

    fn pushCurrentState(self: *Timeline) void {
        self.frames.append(self.allocator, .{
            .cursor = self.current_cursor,
            .hud = self.current_hud,
        }) catch {};
        self.frame_count += 1;
    }

    pub fn writeJson(self: *const Timeline, writer: anytype) !void {
        try writer.writeAll("{");
        try writer.print("\"fps\":{d},\"width\":{d},\"height\":{d},\"zoom\":{d}", .{
            self.fps, self.width, self.height, self.zoom,
        });

        try writer.writeAll(",\"frames\":[");
        for (self.frames.items, 0..) |frame, i| {
            if (i > 0) try writer.writeAll(",");
            try writer.print("{{\"cursor\":{{\"x\":{d},\"y\":{d},\"scale\":{d}}}", .{
                frame.cursor.x, frame.cursor.y, frame.cursor.scale,
            });
            if (frame.hud) |hud| {
                try writer.writeAll(",\"hud\":{\"labels\":[");
                for (hud.labels, 0..) |label, j| {
                    if (j > 0) try writer.writeAll(",");
                    try writeJsonString(writer, label);
                }
                try writer.writeAll("]}");
            } else {
                try writer.writeAll(",\"hud\":null");
            }
            try writer.writeAll("}");
        }
        try writer.writeAll("]");

        try writer.writeAll(",\"events\":[");
        for (self.events.items, 0..) |event, i| {
            if (i > 0) try writer.writeAll(",");
            const type_str: []const u8 = switch (event.event_type) {
                .click => "click",
                .key => "key",
            };
            try writer.print("{{\"type\":\"{s}\",\"timeMs\":{d}}}", .{ type_str, event.time_ms });
        }
        try writer.writeAll("]");

        try writer.writeAll("}");
    }
};

fn writeJsonString(writer: anytype, s: []const u8) !void {
    try writer.writeByte('"');
    for (s) |c| {
        switch (c) {
            '"' => try writer.writeAll("\\\""),
            '\\' => try writer.writeAll("\\\\"),
            '\n' => try writer.writeAll("\\n"),
            '\r' => try writer.writeAll("\\r"),
            '\t' => try writer.writeAll("\\t"),
            else => try writer.writeByte(c),
        }
    }
    try writer.writeByte('"');
}
